### PR TITLE
refactor: merge discovery directly into behaviour and service

### PR DESCRIPTION
**Pull request by ozwaldorf**
- minimal logic used from discovery implemented in behaviour and 
- unified `add_address` logic throughout behavior.rs 
- only perform bootstrap if we're not a bootstrapper and have some nodes
- use verbose paths for QueryId to increase readability as we now have both `kad::QueryId` and `bitswap::QueryId` in behaviour.rs

### CODE REVIEW

**Comment by ozwaldorf** [comment]
Note: I didn't include fixes for the kad test with this, so it's still a false positive as mentioned to refactor tests with another PR separately

File: crates/ursa-network/src/behaviour.rs [hunk 1]
```diff
@@ -159,10 +164,40 @@ impl<P: StoreParams> Behaviour<P> {
                 if relay_client.is_none() {
                     panic!("relay client not instantiated");
                 }
-                dcutr::behaviour::Behaviour::new()
+                Dcutr::new()
             })
             .into();
 
+        let mdns = config
```
Comment (b0xtch) [inline crates/ursa-network/src/behaviour.rs:171]: Revert

File: crates/ursa-network/src/service.rs [hunk 2]
```diff
@@ -196,6 +198,10 @@ pub struct UrsaService<S> {
     _pending_requests: HashMap<RequestId, ResponseChannel<UrsaExchangeResponse>>,
     /// Pending responses.
     pending_responses: HashMap<RequestId, oneshot::Sender<Result<UrsaExchangeResponse>>>,
+    /// Connected peers.
+    peers: HashSet<PeerId>,
```
Comment (b0xtch) [inline crates/ursa-network/src/service.rs:202]:
```rust
    /// Information about connected peers.
    peer_info: HashMap<PeerId, Vec<Multiaddr>>,
```

if are not using peer_info where are we strong the mapping of the `peer_id` and their information

Comment (ozwaldorf) [reply crates/ursa-network/src/service.rs:202]: It's not needed anymore, all the behaviours track this themselves internally and can be accessed at any point with `behaviour.get_peer_addresses()`. This includes addresses from every behaviour internally and you can verify by running a `cargo expand` to view the macro provided implementation

File: crates/ursa-network/src/service.rs [hunk 3]
```diff
@@ -491,19 +492,50 @@ where
...
-            }
-            DiscoveryEvent::Disconnected(peer_id) => {
-                trace!(
-                    "[DiscoveryEvent::PeerDisconnected] - Peer disconnected {:?}",
-                    peer_id
-                );
-                self.emit_event(NetworkEvent::PeerDisconnected(peer_id));
+    pub fn handle_kad(&mut self, event: KademliaEvent) -> Result<()> {
+        match event {
+            KademliaEvent::OutboundQueryProgressed { id, result, .. } => match result {
```
Comment (b0xtch) [inline crates/ursa-network/src/service.rs:497]: Can we handle these kad events

```
GetProviders
GetClosestPeers
```
This is where you can use the peer info

Comment (ozwaldorf) [reply crates/ursa-network/src/service.rs:497]: > GetProviders

This wouldn't ever get generated as we're not using kad's providers or querying for them

> GetClosestPeers

In the current state this event is never generated (it's generated from `behaviour.kad.get_closest_peers()`), but I plan on using + handling it for random kad walks

File: crates/ursa-network/src/service.rs (hunk 2 above)
Comment (b0xtch) [reply crates/ursa-network/src/service.rs:202]: Yeah that make sense given i suggested the move away form having a custom handler still how do we handle custom peer info

Comment (ozwaldorf) [reply crates/ursa-network/src/service.rs:202]: > custom peer info

I didn't see any actual usages of the peer_info struct other than insertions, so I don't think we need to worry for now. What are you anticipating we'd need to store?

Comment (b0xtch) [reply crates/ursa-network/src/service.rs:202]: When we handle custom gossip sub logic with scoring, content routing, and more. You can leave it out, for now, I guess.

File: crates/ursa-network/src/service.rs (hunk 3 above)
Comment (b0xtch) [reply crates/ursa-network/src/service.rs:497]: That make sense

**Review by b0xtch — APPROVED** [review]

File: crates/ursa-network/src/behaviour.rs (hunk 1 above)
Comment (ozwaldorf) [reply crates/ursa-network/src/behaviour.rs]: Done. But code is now inconsistent

image

**Review by b0xtch — APPROVED** [review]
Mint

**Pull request merged**